processor/src/main/java/org/mapstruct/ap/internal/model/common/Parameter.java

```diff
@@ -100,4 +110,36 @@ public boolean equals(Object obj) {
         }
         return true;
     }
+
+    /**
+     * @param parameters the parameters to filter
+     * @return the parameters from the given list that are considered 'source parameters'
+     */
+    public static List<Parameter> getSourceParameters(List<Parameter> parameters) {
+        List<Parameter> sourceParameters = new ArrayList<Parameter>( parameters.size() );
+
+        for ( Parameter parameter : parameters ) {
+            if ( !parameter.isMappingTarget() && !parameter.isTargetType() && !parameter.isMappingContext() ) {
+                sourceParameters.add( parameter );
+            }
+        }
+
+        return sourceParameters;
+    }
+
+    /**
+     * @param parameters the parameters to filter
+     * @return the parameters from the given list that are marked as 'mapping context parameters'
+     */
+    public static List<Parameter> getContextParameters(List<Parameter> parameters) {
+        List<Parameter> contextParameters = new ArrayList<Parameter>( parameters.size() );
+
+        for ( Parameter parameter : parameters ) {
+            if ( parameter.isMappingContext() ) {
+                contextParameters.add( parameter );
+            }
+        }
+
+        return contextParameters;
```

Comment:
I'm a bit more conservative in this area. If we abandon 1.6 and 1.7 we should probably mark that with the transition to MapStruct 2.0. However, I do think that we could profit in the generated code already now. For instance, using `Optional` in the generated code for nested source properties when we detect java8.